### PR TITLE
[PBA-6667] ControlsLockScreenSampleApp - crash when tapped CC-button

### PR DESCRIPTION
**Pull request by oSulyma**
App crashed in `OOClosedCaptionsSelectorBackgroundViewController `because it was nil while adding addChildViewController: (raw 69). It was caused incorrect/uncommon approach for initialization of `OOOoyalaPlayerViewController` (in storyboard, but without calling flow in methods with Initialization). So I did it as i *BasicPlaybackSampleApp*.
Also player used to miss reference to embedTokenGenerator because weak ref.
Fixed.

### CODE REVIEW

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 1]
```diff
@@ -123,23 +165,24 @@ class PlayerViewController: OOOoyalaPlayerViewController {
...
+    if let player = ooyalaPlayerVC.player {
       player.seek(player.playheadTime() - sender.interval)
       player.play()
       updatePlayingInfoCenter()
     }
   }
   
   @objc func skipForwardCommand(_ sender: MPSkipIntervalCommandEvent) {
-    if player != nil {
+    if let player = ooyalaPlayerVC.player {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Prefer `guard let` over `if let` if possible

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 2]
```diff
@@ -123,23 +165,24 @@ class PlayerViewController: OOOoyalaPlayerViewController {
...
       player.seek(player.playheadTime() + sender.interval)
       player.play()
       updatePlayingInfoCenter()
     }
   }
   
   @objc func changePlaybackPositionCommand(_ sender: MPChangePlaybackPositionCommandEvent) {
-    if player != nil {
+    if let player = ooyalaPlayerVC.player {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Prefer `guard let` over `if let` if possible

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 3]
```diff
@@ -41,29 +73,38 @@ class PlayerViewController: OOOoyalaPlayerViewController {
...
-    setupCommandCenter()
-    setupPlayingInfoCenter()
-  }
-  
-  override func viewDidAppear(_ animated: Bool) {
-    super.viewDidAppear(animated)
-    player.play()
+
+    ooyalaPlayerVC = OOOoyalaPlayerViewController(player: player)
+    // Load the video, step 1/2
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Remove commented code

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 4]
```diff
@@ -9,23 +9,55 @@ import UIKit
...
+    vc.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+    self.addChild(vc)
+    view.addSubview(vc.view)
+  }
+  
+  private func setupOoyalaPlayerStuff() {
+    
+    var player: OOOoyalaPlayer!
+    
     if option.embedTokenGenerator != nil {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Prefer `guard let` over `if let` if possible

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 5]
```diff
@@ -9,23 +9,55 @@ import UIKit
...
+  private func addPlayerViewController(vc: UIViewController) {
+    
+    vc.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+    self.addChild(vc)
+    view.addSubview(vc.view)
+  }
+  
+  private func setupOoyalaPlayerStuff() {
+    
+    var player: OOOoyalaPlayer!
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Does it really need to be an implicit unwrapped optional?

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 6]
```diff
@@ -9,23 +9,55 @@ import UIKit
...
+  override func viewDidAppear(_ animated: Bool) {
+    super.viewDidAppear(animated)
+    
+    // Load the video, step 2/2
+    ooyalaPlayerVC.player.play()
+  }
+  
+  // MARK: - Private methods
+  private func addPlayerViewController(vc: UIViewController) {
+
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Remove extra line

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 7]
```diff
@@ -9,23 +9,55 @@ import UIKit
...
+class PlayerViewController: UIViewController {
   
+  // MARK: - Private properties
   // properties for the video
   public var option: PlayerSelectionOption!
   
   // properties required for a Fairplay asset
   private var apiKey: String?
   private var apiSecret: String?
+  private var ooyalaPlayerVC: OOOoyalaPlayerViewController!
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Does it really need to be an implicit unwrapped optional?

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 8]
```diff
@@ -100,18 +141,19 @@ class PlayerViewController: OOOoyalaPlayerViewController {
   }
   
   // Updates the time labels.
-  func updatePlayingInfoCenter() {
+  private func updatePlayingInfoCenter() {
     let playingInfoCenter = MPNowPlayingInfoCenter.default()
-    if var displayInfo = playingInfoCenter.nowPlayingInfo {
+    if var displayInfo = playingInfoCenter.nowPlayingInfo, let player = ooyalaPlayerVC.player {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Prefer `guard let` over `if let` if possible

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 9]
```diff
@@ -123,23 +165,24 @@ class PlayerViewController: OOOoyalaPlayerViewController {
   }
   
   @objc func skipBackwardCommand(_ sender: MPSkipIntervalCommandEvent) {
-    if player != nil {
+    
+    if let player = ooyalaPlayerVC.player {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Prefer `guard let` over `if let` if possible

**Review by maxkupetskii — CHANGES_REQUESTED** [review]

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 7 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Description of initialization method says that result is not-optional, that's why I used non-optional declaration. 
_Declaration: init!(player: OOOoyalaPlayer!)_

Should I avoid this even when init is not-optional?

**Review by DmitriiMaslov — COMMENTED** [review]
Guys, lets point it to dev. I don't want to include anything else into candidate. We should go to release today.

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 1 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 2 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 4 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 5 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 6 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 7 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Changed to optional

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 8 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

**Comment by oSulyma** [comment]
@maxkupetskii  I've done all changes, please, take a look again.

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 9 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 3 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 10]
```diff
@@ -100,55 +140,60 @@ class PlayerViewController: OOOoyalaPlayerViewController {
...
-      else {
-        player.play()
-      }
-      updatePlayingInfoCenter()
+    guard let player = ooyalaPlayerVC?.player else { return }
+    
+    if player.isPlaying() {
+      player.pause()
+    }
+    else {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Use `if ... else` with following structure
```
if ... {
  ...
  } else {
  ...
}
```

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 11]
```diff
@@ -9,27 +9,57 @@ import UIKit
...
+  
+  private func setupOoyalaPlayerStuff() {
+    var player: OOOoyalaPlayer?
+    
+    if let embedTokenGenerator = option.embedTokenGenerator {
+      if let basicEmbedTokenGenerator = embedTokenGenerator as? BasicEmbedTokenGenerator {
+        apiKey = basicEmbedTokenGenerator.apiKey
+        apiSecret = basicEmbedTokenGenerator.apiSecret
       }
       else {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Use `if ... else` with following appearance
```
if ... {
  ...
  } else {
  ...
}
```

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 12]
```diff
@@ -41,29 +71,39 @@ class PlayerViewController: OOOoyalaPlayerViewController {
       // This is not how this should be implemented in production - In production, you should implement your own OOSecureURLGenerator
       // which contacts a server of your own, which will help sign the url with the appropriate API Key and Secret
       options.secureURLGenerator = OOEmbeddedSecureURLGenerator(apiKey: apiKey, secret: apiSecret)!
-      player = OOOoyalaPlayer(pcode: option.pcode, domain: option.domain, embedTokenGenerator: option.embedTokenGenerator!, options: options)
+      player = OOOoyalaPlayer(pcode: option.pcode,
+                              domain: option.domain,
+                              embedTokenGenerator: embedTokenGenerator,
+                              options: options)
     }
     else {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Use `if ... else` with following appearance
```
if ... {
  ...
  } else {
  ...
}
```

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift [hunk 13]
```diff
@@ -100,55 +140,60 @@ class PlayerViewController: OOOoyalaPlayerViewController {
...
   // Updates the time labels.
-  func updatePlayingInfoCenter() {
+  private func updatePlayingInfoCenter() {
     let playingInfoCenter = MPNowPlayingInfoCenter.default()
-    if var displayInfo = playingInfoCenter.nowPlayingInfo {
-      displayInfo[MPNowPlayingInfoPropertyPlaybackRate] = player.isPlaying() ? 1 : 0
-      displayInfo[MPMediaItemPropertyPlaybackDuration] = player.duration()
-      displayInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = player.playheadTime()
-      playingInfoCenter.nowPlayingInfo = displayInfo
+    guard var displayInfo = playingInfoCenter.nowPlayingInfo, let player = ooyalaPlayerVC?.player else {
```
Comment (maxkupetskii) [inline PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Please, align `var displayInfo` and `let player` on 2 lines, i.e.
```
guard var displayInfo = ...,
      let player = ...
```

**Review by maxkupetskii — CHANGES_REQUESTED** [review]

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 10 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 11 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 12 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

File: PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift (hunk 13 above)
Comment (oSulyma) [reply PlaybackLab/ControlsLockScreenSampleApp/ControlsLockScreenSampleApp/Views/PlayerViewController.swift]: Done

**Review by maxkupetskii — APPROVED** [review]

**Review by cacel — APPROVED** [review]
LGTM

**Pull request merged**